src/tensorwaves/model.py

```diff
@@ -77,6 +143,135 @@ def update_parameters(
+            self.__not_fixed_parameters
+        )
+
+    def __find_constant_subexpressions(self, expr: sp.Expr) -> bool:
+        if not expr.args:
+            if (
+                isinstance(expr, sp.Symbol)
+                and expr.name not in self.__constant_symbols
+            ):
+                return False
+            return True
+
+        is_constant = True
+        temp_constant_sub_expression = []
+        for arg in expr.args:
+            if self.__find_constant_subexpressions(arg):
+                if arg.args:
+                    temp_constant_sub_expression.append(arg)
+            else:
+                is_constant = False
+
+        if not is_constant:
+            for sub_expr in temp_constant_sub_expression:
+                placeholder = sp.Symbol(f"cached[{str(sub_expr)}]")
+                self.__constant_sub_expressions[placeholder] = sub_expr
+        return is_constant
+
+    def __replace_constant_sub_expressions(
+        self, expression: sp.Expr
+    ) -> sp.Expr:
+        new_expression = copy.deepcopy(expression)
+        return new_expression.xreplace(
+            {v: k for k, v in self.__constant_sub_expressions.items()}
+        )
+
+    def lambdify(self, backend: Union[str, tuple, dict]) -> Callable:
+        input_symbols = tuple(self.__expression.free_symbols)
+        lambdified_model = _sympy_lambdify(
+            input_symbols,
+            self.__expression,
+            backend=backend,
+        )
+        constant_input_storage = {}
+        for placeholder, sub_expr in self.__constant_sub_expressions.items():
+            temp_lambdify = _sympy_lambdify(
+                tuple(sub_expr.free_symbols), sub_expr, backend
+            )
+            free_symbol_names = {x.name for x in sub_expr.free_symbols}
+            constant_input_storage[placeholder.name] = temp_lambdify(
+                *(self.__fix_inputs[k] for k in free_symbol_names)
+            )
+
+        input_args: list = []
+        non_fixed_arg_positions = list(range(0, len(self.argument_order)))
+
+        for input_arg in input_symbols:
+            if input_arg in self.__argument_order:
+                non_fixed_arg_positions[
+                    self.__argument_order.index(input_arg)
+                ] = len(input_args)
+                input_args.append(0.0)
+            elif input_arg.name in self.__fix_inputs:
+                input_args.append(self.__fix_inputs[input_arg.name])
+            else:
+                input_args.append(constant_input_storage[input_arg.name])
+
+        def update_args(*args: Tuple[Any, ...]) -> None:
+            for i, x in enumerate(args):
+                input_args[non_fixed_arg_positions[i]] = x
+
+        def wrapper(*args: Tuple[Any, ...]) -> Any:
+            update_args(*args)
+            return lambdified_model(*input_args)
+
+        return wrapper
+
+    def performance_optimize(self, fix_inputs: DataSample) -> "Model":
+        return NotImplemented
```

Comment:
Shouldn't this method return the same as `SympyModel.performance_optimize()`? So returning another `_ConstantSubExpressionSympyModel`?

Reply:
I want this returned instance to implement the Model Interface